keps/sig-api-machinery/20190607-owner-ref.md

```diff
@@ -0,0 +1,128 @@
+every resource in the cluster to find a list of potential resources (URLs more or less).  This approximates user intent
+and mostly works, but it is inefficient and unnecessarily vulnerable to network outages.
+
+This KEP proposes adding an optional `resource` field to OwnerReferences.  If provided, this will be the authoritative value and
+no lookup or approximation of user intent will be required.  If not provided, the `kind` will be used as it is today.
+If both are provided, old clusters continue to work and new clusters will be more efficient and resilient.
+ 
+```yaml
+ownerReferences:
+- apiVersion: apps/v1
+  kind: DaemonSet
+  resource: daemonsets
+```
+
+## Motivation
+
+* Allow precision.  `kind` is a one to many relationship that makes the exact object reference inefficient or impossible to determine.
+* Remove remote calls.  Because a `kind` to `resource` mapping requires discovery, it is vulnerable to network segmentation
+ 
+### Goals
+
+* Improve reliability. 
+* Improve efficiency.
+* Improve predictability.
+* Avoid disrupting existing clients.
+
+### Non-Goals
+
+* Improving discovery.  No amount of improving speed can fix predictability and we're already doing what caching is practical.
+* Improve rest mapping.  No amount of feature enhancement can fix predictability.
+
+## Proposal
+
+Add a new, optional field to `OwnerReferences` that represents the `resource`.
+
+```go
+type OwnerReference struct {
+	// Kind of the referent.
+	Kind string `json:"kind" protobuf:"bytes,1,opt,name=kind"`
+	// Resource of the referent.
+	// +optional
+	Resource string `json:"resource,omitempty" protobuf:"bytes,8,opt,name=resource"`
+}
+```
+
+The only impact is in the case of a new cluster with both fields set.  In this case, the new cluster does not have to 
+perform any lookups and will trust the client's specification of `resource`.
+
+Existing clients are not required to make this change, but it will improve reliability.  If a client sets the value improperly,
+then the behavior is congruent to behavior with a mis-set `kind`.  If the kind (resource) is valid, but incorrect 
+GC collects the object immediately.  If the kind (resource) is invalid, GC reports the error and waits.  
+
+
+### Behavior of new clusters
+1. `resource` set and `kind` both set: only `resource` is honored and no mapping required  
+2. `resource` set and `kind` not set: validation failure.  Allowing this would break old clusters and clients.  
+3. `resource` not set and `kind` set: `kind` is honored, exactly as before
+4. `resource` not set and `kind` not set: validation failure
+  
+### Behavior of old clusters
+1. `resource` set and `kind` both set: only `kind` is honored, exactly as before  
+2. `resource` set and `kind` not set: validation failure  
+3. `resource` not set and `kind` set: `kind` is honored, exactly as before
+4. `resource` not set and `kind` not set: validation failure
+
```

Comment:
also need to consider behavior of old clients that drop the new field, and the affect on things like `isChangingOwnerReference` in `gc_admission.go`

Reply:
> also need to consider behavior of old clients that drop the new field, and the effect on things like `isChangingOwnerReference` in `gc_admission.go`

I will update.  I think it's as simple as considering the removal of the `resource` field with not other mutation to be a non-change.  Presuming a well-formed initial reference (the behavior of GC more or less require this), the update would be allowed and the controller would behave as expected because the map is built out of all resources.

Reply:
hmm... that would mean I could do the following as a user without delete permission on Nodes:

1. create an object with:
```json
ownerRef: {
  kind:"Node",
  resource:"configmaps",
  name:"node1",
  uid:"uid-of-node1",
  blockOwnerDeletion: true
}
```

2. the gc_admission plugin would authorize me based on delete permission on configmaps

3. immediately (racing GC) patch the object to remove resource:
```json
ownerRef: {
  kind:"Node",
  name:"node1",
  uid:"uid-of-node1",
  blockOwnerDeletion: true
}
```

If you don't reauthorize, you allow blocking deletion of the node

Reply:
> If you don't reauthorize, you allow blocking deletion of the node
> ```

I'm fine with checking to see if a mapping confirms equivalence.

Reply:
Is this fixed in the text?